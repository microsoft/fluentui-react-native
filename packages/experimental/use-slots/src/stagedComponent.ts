import * as React from 'react';

/**
 * This is a pattern of rendering where a functional component can be executed in two stages rather than in a single pass.
 *
 * The pattern looks like:
 * (props) => {
 *   // handle props
 *   // call hooks, remember these can't be conditional
 *   // build styles and props to pass to child components
 *
 *   return (additionalProps) => {
 *     // return the actual element tree, this includes conditional branching or rendering
 *     // mixin additional props, props which require logic should be required in phase 1.
 *   };
 * }
 */

export type StagedRender<TProps> = (props: TProps) => React.FunctionComponent<TProps>;

/**
 * A composable function may have a two stage render function as an attached property. This allows the function to work
 * in all the standard react flows, but allows for pulling out the staged render when components understand it.
 */
export type ComposableFunction<TProps> = React.FunctionComponent<TProps> & { _staged?: StagedRender<TProps> };

/**
 * Take a staged render function and make a real component out of it
 *
 * @param staged - staged render function to wrap into a staged component
 * @param memo - optional flag to enable wrapping the created component in a React.memo HOC
 */
export function stagedComponent<TProps>(staged: StagedRender<TProps>, memo?: boolean): ComposableFunction<TProps> {
  const component: ComposableFunction<TProps> = memo
    ? ((React.memo((props: TProps) => staged(props)({} as TProps)) as unknown) as ComposableFunction<TProps>)
    : (props: TProps) => staged(props)({} as TProps);
  component._staged = staged;
  return component;
}
