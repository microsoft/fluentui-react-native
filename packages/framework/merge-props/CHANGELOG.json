{
  "name": "@fluentui-react-native/merge-props",
  "entries": [
    {
      "date": "Fri, 25 Sep 2020 19:21:43 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.3",
      "version": "0.1.3",
      "comments": {
        "patch": [
          {
            "comment": "Update react-native-win32 versions - enable logbox",
            "author": "acoates-ms@noreply.github.com",
            "commit": "a95ab1b66822cceacc74e24094f85c6917f58086",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Wed, 23 Sep 2020 18:31:48 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.2",
      "version": "0.1.2",
      "comments": {
        "patch": [
          {
            "comment": "start publishing src to fix customer source maps",
            "author": "jasonmo@microsoft.com",
            "commit": "b3d7b5ba710bc581a6f4db91c91e4ffc185fc427",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Fri, 11 Sep 2020 20:25:42 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "Add samples and some README text to use-slots and use-styling packages",
            "author": "jasonmo@microsoft.com",
            "commit": "2566347850eff4ed67bfad2a85134481e088cba6",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Fri, 28 Aug 2020 05:03:51 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "only allow menu container to focus on mount",
            "author": "lehon@microsoft.com",
            "commit": "9c93388527af2ea3ba60336c56a3da0ce65e3688",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Thu, 20 Aug 2020 00:14:01 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "split mergeProps and mergeStyles out of foundation-settings",
            "author": "jasonmo@microsoft.com",
            "commit": "f75e45ede85b8ebc6cd11cc66fbb1baedaea39b9",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    }
  ]
}
