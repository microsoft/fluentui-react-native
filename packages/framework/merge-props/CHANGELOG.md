# Change Log - @fluentui-react-native/merge-props

This log was last generated on Fri, 25 Sep 2020 19:21:43 GMT and should not be manually modified.

<!-- Start content -->

## 0.1.3

Fri, 25 Sep 2020 19:21:43 GMT

### Patches

- Update react-native-win32 versions - enable logbox (acoates-ms@noreply.github.com)

## 0.1.2

Wed, 23 Sep 2020 18:31:48 GMT

### Patches

- start publishing src to fix customer source maps (jasonmo@microsoft.com)

## 0.1.1

Thu, 20 Aug 2020 00:14:01 GMT

### Patches

- split mergeProps and mergeStyles out of foundation-settings (jasonmo@microsoft.com)
