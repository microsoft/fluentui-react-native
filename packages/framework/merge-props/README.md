# `@fluentui-react-native/merge-props`

This package is now a pass-through package which references `@fluentui-react-native/framework-base` for backwards compatibility.

The current documentation can be found [here](../../framework-base/src/merge-props/README.md).
