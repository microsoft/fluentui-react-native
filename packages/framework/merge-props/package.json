{
  "name": "@fluentui-react-native/merge-props",
  "version": "0.8.0",
  "description": "Utility for merging props with styles and caching style combinations",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/fluentui-react-native",
    "directory": "packages/framework/merge-props"
  },
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "test": "fluentui-scripts jest",
    "lint": "fluentui-scripts eslint",
    "start": "fluentui-scripts dev",
    "start-test": "fluentui-scripts jest-watch",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@fluentui-react-native/immutable-merge": "^1.2.1",
    "@fluentui-react-native/memo-cache": "^1.3.1",
    "tslib": "^2.3.1"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "^0.1.1",
    "@fluentui-react-native/scripts": "^0.1.1",
    "@react-native/metro-config": "^0.72.0",
    "@types/jest": "^29.0.0",
    "react": "18.2.0",
    "react-native": "^0.72.0"
  },
  "peerDependencies": {
    "react": "18.2.0",
    "react-native": "^0.72.0"
  },
  "rnx-kit": {
    "kitType": "library",
    "alignDeps": {
      "presets": [
        "microsoft/react-native"
      ],
      "requirements": [
        "react-native@0.72"
      ],
      "capabilities": [
        "core",
        "core-android",
        "core-ios"
      ]
    }
  }
}
