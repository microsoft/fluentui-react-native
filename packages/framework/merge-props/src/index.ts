export { StyleProp } from './mergeStyles.types';
export { mergeStyles } from './mergeStyles';
export { mergeProps } from './mergeProps';
