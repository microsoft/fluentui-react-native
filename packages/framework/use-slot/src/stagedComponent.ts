import * as React from 'react';

/**
 * The final rendering of the props in a staged render. This is the function component signature that matches that of
 * React.createElement, children (if present) will be part of the variable args at the end.
 */
export type FinalRender<TProps> = (props: TProps, ...children: React.ReactNode[]) => JSX.Element | null;

/**
 * This is a pattern of rendering where a functional component can be executed in two stages rather than in a single pass.
 *
 * The pattern looks like:
 * (props) => {
 *   // handle props
 *   // call hooks, remember these can't be conditional
 *   // build styles and props to pass to child components
 *
 *   return (additionalProps, ...children) => {
 *     // return the actual element tree, this includes conditional branching or rendering
 *     // mixin additional props, props which require logic should be required in phase 1.
 *
 *     // NOTE: This is where children will show up
 *   };
 * }
 */

export type StagedRender<TProps> = (props: TProps) => FinalRender<TProps>;

/**
 * A composable function may have a two stage render function as an attached property. This allows the function to work
 * in all the standard react flows, but allows for pulling out the staged render when components understand it.
 */
export type ComposableFunction<TProps> = React.FunctionComponent<TProps> & { _staged?: StagedRender<TProps> };

function asArray<T>(val: T | T[]): T[] {
  return Array.isArray(val) ? val : [val];
}

/**
 * Take a staged render function and make a real component out of it
 *
 * @param staged - staged render function to wrap into a staged component
 * @param memo - optional flag to enable wrapping the created component in a React.memo HOC
 */
export function stagedComponent<TProps>(staged: StagedRender<TProps>, memo?: boolean): ComposableFunction<TProps> {
  const test = (props: React.PropsWithChildren<TProps>) => {
    const { children, ...rest } = props;
    return staged(rest as TProps)({} as React.PropsWithChildren<TProps>, ...asArray(children));
  };
  const stagedComponent = memo ? React.memo(test) : test;
  Object.assign(stagedComponent, { _staged: staged });
  return stagedComponent as ComposableFunction<TProps>;
}
