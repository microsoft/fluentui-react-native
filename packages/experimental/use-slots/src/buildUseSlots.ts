import * as React from 'react';
import { SlotFn, NativeReactType } from './renderSlot';
import { mergeProps } from '@uifabricshared/foundation-settings';
import { ComposableFunction, StagedRender } from './stagedComponent';

export type Slots<TSlotProps> = { [K in keyof TSlotProps]: SlotFn<TSlotProps[K]> };

export type UseSlotOptions<TSlotProps> = {
  slots: { [K in keyof TSlotProps]: NativeReactType | ComposableFunction<TSlotProps[K]> };
  filters?: { [K in keyof TSlotProps]?: (propName: string) => boolean };
  useStyling?: TSlotProps | GetSlotProps<TSlotProps>;
};

export type GetSlotProps<TSlotProps> = (...args: any[]) => TSlotProps;

export type UseSlotsBase<TSlotProps> = (...args: any[]) => Slots<TSlotProps>;

type CachedState<TSlotProps> = {
  /**
   * closures for each slot, created once and cached in the state
   */
  slots: Slots<TSlotProps>;

  /**
   * results of pre-render phase, either props to pass to render, or a render function returned from phase one
   */
  results: { [K in keyof TSlotProps]?: React.FunctionComponent<TSlotProps[K]> | TSlotProps[K] };
};

function internalRender<TProps>(
  slot: NativeReactType | ComposableFunction<TProps>,
  result: React.FunctionComponent<TProps> | TProps,
  extraProps: TProps,
  filter: (propsName: string) => boolean,
  children: React.ReactNode[]
): JSX.Element | null {
  let props: TProps = typeof result === 'function' ? extraProps : mergeProps(result, extraProps);
  const propsToRemove = filter ? Object.keys(props).filter(key => !filter(key)) : undefined;
  if (propsToRemove?.length > 0) {
    props = mergeProps(props, (Object.assign({}, ...propsToRemove.map(prop => ({ [prop]: undefined }))) as unknown) as TProps);
  }
  return typeof result === 'function' ? (result as Function)(props, ...children) : React.createElement(slot, props, ...children);
}

function getStagedRender<TProps>(slot: NativeReactType | ComposableFunction<TProps>): StagedRender<TProps> | undefined {
  return (typeof slot === 'function' && (slot as ComposableFunction<TProps>)._staged) || undefined;
}

function buildSlotFunctions<TSlotProps>(
  slots: UseSlotOptions<TSlotProps>['slots'],
  filters?: UseSlotOptions<TSlotProps>['filters']
): CachedState<TSlotProps> {
  const info: CachedState<TSlotProps> = { slots: {}, results: {} } as CachedState<TSlotProps>;
  Object.keys(slots).forEach(slot => {
    info.slots[slot] = (props: TSlotProps[keyof TSlotProps], ...children: React.ReactNode[]) => {
      return internalRender<TSlotProps[keyof TSlotProps]>(slots[slot], info.results[slot], props, (filters && filters[slot]) || undefined, children);
    };
    info.slots[slot]._canCompose = true;
  });
  return info as CachedState<TSlotProps>;
}

export function buildUseSlots<TSlotProps>(options: UseSlotOptions<TSlotProps>): UseSlotsBase<TSlotProps> {
  const { slots, filters, useStyling } = options;
  return (...args: any[]) => {
    // build up a set of slots closures and store them in props
    const state = React.useMemo(() => buildSlotFunctions<TSlotProps>(slots, filters), []);

    // get the baseline slot props to render with the slots
    const slotProps: TSlotProps = typeof useStyling === 'function' ? (useStyling as Function)(...args) : ((useStyling || {}) as TSlotProps);

    // for each slot go through and either cache the slot props or call part one render if it is staged
    Object.keys(slots).forEach(slotName => {
      const props = slotProps[slotName];
      const staged = getStagedRender(slots[slotName]);
      state.results[slotName] = staged ? staged(props) : props;
    });

    // return the prebuilt closures, these will have internal references to state.results
    return state.slots;
  };
}
