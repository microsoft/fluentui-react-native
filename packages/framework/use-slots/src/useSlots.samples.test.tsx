/** @jsx withSlots */
import { withSlots } from './withSlots';
import { stagedComponent } from './stagedComponent';
import { mergeProps } from '@fluentui-react-native/merge-props';
import { buildUseSlots } from './buildUseSlots';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import { CSSProperties } from 'react';

// types for web
type TextProps = { style?: CSSProperties };
type ViewProps = { style?: CSSProperties };
type ViewStyle = CSSProperties;
type TextStyle = CSSProperties;

/**
 * This file contains samples and description to help explain what the useSlots hook does and why it is useful
 * for building components.
 */

describe('useSlots sample code test suite', () => {
  /**
   * The first mechanism to understand is the stagedComponent mechanic. This allows a component to be written, separating
   * hook calls and component rendering. This allows it to be safely called as a function by a higher order component, even conditionally.
   */

  /**
   * Example #1: Single level simple component ----------------------------------------
   *
   * First we are going to create a wrapped text component that bolds all text. One component will be authored as a staged
   * component and one as a regular component.
   */

  const boldBaseProps: TextProps = { style: { fontWeight: 900 } };

  /**
   * First create the bold text in the standard way. This is just a function component.
   */
  const BoldTextStandard: React.FunctionComponent<TextProps> = (props: React.PropsWithChildren<TextProps>) => {
    /**
     * Pick out the children to pass them on to the child Text element
     */
    const { children, ...rest } = props;

    /**
     * Now render the text, merging the baseProps with the style updates with the rest param. Note that this leverages the fact
     * that mergeProps will reliably produce style objects with the same reference, given the same inputs.
     */
    return <span {...mergeProps(boldBaseProps, rest)}>{children}</span>;
  };
  BoldTextStandard.displayName = 'BoldTextStandard';

  /**
   * To write the same component using the staged pattern is only slightly more complex. The pattern involves splitting the component rendering into
   * two parts and executing any hooks in the first part.
   *
   * The stagedComponent function takes an input function of this form and wraps it in a function component that react knows how to render
   */
  const BoldTextStaged = stagedComponent((props: TextProps) => {
    /**
     * This section would be where hook/styling code would go, props here would include everything coming in from the base react tree with the
     * exception of children, which will be passed in stage 2.
     */
    return (extra: TextProps, children: React.ReactNode) => {
      /**
       * extra are additional props that may be filled in by a higher order component. They should not include styling and are only props the
       * enclosing component are passing to the JSX elements
       */
      return <span {...mergeProps(boldBaseProps, props, extra)}>{children}</span>;
    };
  });
  BoldTextStaged.displayName = 'BoldTextStaged';

  /**
   * The demos of the code use enzyme with JSDom to show the full tree. This has the side effect of doubling up primitive elements in the output
   * JSON. This is an issue with rendering react-native with enzyme but in real usage the nodes only render once.
   */
  it('renders sample 1 - the two types of basic bold text components', () => {
    const styleToMerge: TextStyle = { color: 'black' };

    /**
     * First render the staged component. This invokes the wrapper that was built by the stagedComponent function
     */
    const wrapper = mount(
      <div>
        <BoldTextStaged style={styleToMerge}>Staged component at one level</BoldTextStaged>
        <BoldTextStandard style={styleToMerge}>Standard component of a single level</BoldTextStandard>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Example #2 - Simple component containing another simple component -------------------------------------
   *
   * Next we will build a layer on top of the previously authored components to turn the bold text components into header components. This is
   * to illustrate the way in which components can be commonly built on top of other simpler components.
   */
  const headerBaseProps: TextProps = { style: { fontSize: 20 } };

  /**
   * The standard way of doing things is a repeat of what happens above. Grab the children, pass them on, merge the rest of the props with
   * base props.
   *
   * This again leverages style merging via mergeProps to avoid changing the references of the style objects on every render
   */
  const HeaderStandard: React.FunctionComponent<TextProps> = (props) => {
    const { children, ...rest } = props;
    return <BoldTextStandard {...mergeProps(headerBaseProps, rest)}>{children}</BoldTextStandard>;
  };
  HeaderStandard.displayName = 'HeaderStandard';

  /**
   * To consume the staged component we'll use the use slots hook builder. This allows easy consumption of staged components (or standard components)
   * This will be described in more detail further on but in this case the component has a single child component, so it only has one slot that we
   * will call 'text'
   *
   * This should be built once, and consumed by the component, not built on the fly inside
   */
  const useHeaderSlots = buildUseSlots({ slots: { text: BoldTextStaged } });

  /**
   * Now author the staged component using the slot hook
   */
  const HeaderStaged = stagedComponent((props: TextProps) => {
    /**
     * Call the slots hook (or any hook) outside of the inner closure. The useSlots hook will return an object with each slot as a renderable
     * function. The hooks for sub-components will be called as part of this call. Props passed in at this point will be the props that appear
     * in outer part of the staged component. (For this example `props`)
     *
     * Note that while we are passing in props, in the simple usage case it isn't used and could be omitted if desired
     * */
    const BoldText = useHeaderSlots(props).text;

    /** Now the inner closure, pretty much the same as before */
    return (extra: TextProps, children: React.ReactNode) => {
      /**
       * Instead of rendering the <BoldTextStageed> component directly we render using the slot. If this is a staged component it will call the
       * inner closure directly, without going through createElement. Entries passed into the JSX, including children, are what appear in the
       * props of the inner closure. (In this example `extra`)
       *
       * NOTE: this requires using the withSlots helper via the jsx directive. This knows how to pick apart the entries and just call the second
       * part of the function
       */
      return <BoldText {...mergeProps(headerBaseProps, props, extra)}>{children}</BoldText>;
    };
  });
  HeaderStaged.displayName = 'HeaderStaged';

  /**
   * Look at the snapshots to compare the rendered output. The staged component will skip the intermediate levels of the react hieararchy while
   * still rendering to the correct primitives.
   */
  it('renders sample 2 = the two types of two level header components', () => {
    const styleToMerge: TextStyle = { color: 'black' };

    /**
     * First render the staged component. This invokes the wrapper that was built by the stagedComponent function
     */
    const wrapper = mount(
      <div>
        <HeaderStaged style={styleToMerge}>Staged component with two levels</HeaderStaged>
        <HeaderStandard style={styleToMerge}>Standard component with two levels</HeaderStandard>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Example #3: Complex components built using useSlots --------------------------------------------------------------------
   *
   * We'll build a component that shows two labels, a header and a caption, embedded inside a view
   */

  type HeaderWithCaptionProps = ViewProps & { headerColor?: string; captionColor?: string; captionText?: string };

  /** standard props for the container */
  const containerProps: ViewProps = { style: { display: 'flex', flexDirection: 'column' } };

  /**
   * add a quick cache to ensure that we don't thrash the styles. This is a danger any time a value from a style is added as
   * a prop on a component
   */
  const colorProps = {};
  const getColorProps = (value?: string) => {
    if (value !== undefined) {
      colorProps[value] = colorProps[value] || { style: { color: value } };
      return colorProps[value];
    }
    return {};
  };

  /**
   * now just create the component like a standard react functional component
   */
  const CaptionedHeaderStandard: React.FunctionComponent<HeaderWithCaptionProps> = (props) => {
    const { headerColor, captionColor, captionText, children, ...rest } = props;
    const headerColorProps = getColorProps(headerColor);
    const captionColorProps = getColorProps(captionColor);
    return (
      <div {...mergeProps(containerProps, rest)}>
        <HeaderStandard {...headerColorProps}>{children}</HeaderStandard>
        {captionText && <BoldTextStandard {...captionColorProps}>{captionText}</BoldTextStandard>}
      </div>
    );
  };
  CaptionedHeaderStandard.displayName = `CaptionedHeaderStandard';`;

  /**
   * now build the same component using slots hook. This will also add use of the style injection pattern
   */
  const useCaptionedHeaderSlots = buildUseSlots({
    /** Slots are just like above, this component will have three sub-components */
    slots: {
      container: 'div',
      header: HeaderStaged,
      caption: BoldTextStaged,
    },
    /** useStyling is an optional function that turns props into props for the sub-components */
    useStyling: (props: HeaderWithCaptionProps) => ({
      container: containerProps,
      header: getColorProps(props.headerColor),
      caption: getColorProps(props.captionColor),
    }),
  });

  /** a mask to clear props that we don't want to pass to the inner view */
  const clearCustomProps = { headerColor: undefined, captionColor: undefined };

  /**
   * now use the hook to implement it as a staged component
   */
  const CaptionedHeaderStaged = stagedComponent<HeaderWithCaptionProps>((props) => {
    // At the point where this is called the slots are initialized with the initial prop values from useStyling above
    const Slots = useCaptionedHeaderSlots(props);
    return (extra: HeaderWithCaptionProps, children: React.ReactNode) => {
      // merge the props together, picking out the caption text and clearing any custom values we don't want forwarded to the view
      const { captionText, ...rest } = mergeProps(props, extra, clearCustomProps);

      // now render using the slots. Any values passed in via JSX will be merged with values from the slot hook above
      return (
        <Slots.container {...rest}>
          <Slots.header>{children}</Slots.header>
          {captionText && <Slots.caption>{captionText}</Slots.caption>}
        </Slots.container>
      );
    };
  });
  CaptionedHeaderStaged.displayName = 'CaptionedHeaderStaged';

  /**
   * Render to enzyme snapshots
   */
  it('renders sample 3 - the two types of higher order header components', () => {
    const styleToMerge: ViewStyle = { backgroundColor: 'gray', borderColor: 'purple', borderWidth: 1 };

    /**
     * Render the two sets of components. Note in the snapshots how the render tree layers for the standard approach are starting
     * to add up.
     */
    const wrapper = mount(
      <div>
        <span>--- SIMPLE USAGE COMPARISON ---</span>
        <CaptionedHeaderStandard style={styleToMerge}>Standard HOC</CaptionedHeaderStandard>
        <CaptionedHeaderStaged style={styleToMerge}>Staged HOC</CaptionedHeaderStaged>
        <span>--- COMPARISON WITH CAPTIONS ---</span>
        <CaptionedHeaderStandard style={styleToMerge} captionText="Caption text">
          Standard HOC with Caption
        </CaptionedHeaderStandard>
        <CaptionedHeaderStaged style={styleToMerge} captionText="Caption text">
          Staged HOC with Caption
        </CaptionedHeaderStaged>
        <span>--- COMPARISON WITH CAPTIONS AND CUSTOMIZATIONS ---</span>
        <CaptionedHeaderStandard style={styleToMerge} captionText="Caption text" captionColor="yellow" headerColor="red">
          Standard HOC with caption and customizations
        </CaptionedHeaderStandard>
        <CaptionedHeaderStaged style={styleToMerge} captionText="Caption text" captionColor="yellow" headerColor="red">
          Staged HOC with caption and customizations
        </CaptionedHeaderStaged>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
