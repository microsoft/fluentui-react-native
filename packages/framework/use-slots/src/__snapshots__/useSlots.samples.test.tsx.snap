// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSlots sample code test suite renders sample 1 - the two types of basic bold text components 1`] = `
<div>
  <span
    style={
      {
        "color": "black",
        "fontWeight": 900,
      }
    }
  >
    Staged component at one level
  </span>
  <span
    style={
      {
        "color": "black",
        "fontWeight": 900,
      }
    }
  >
    Standard component of a single level
  </span>
</div>
`;

exports[`useSlots sample code test suite renders sample 2 = the two types of two level header components 1`] = `
<div>
  <span
    style={
      {
        "color": "black",
        "fontSize": 20,
        "fontWeight": 900,
      }
    }
  >
    Staged component with two levels
  </span>
  <span
    style={
      {
        "color": "black",
        "fontSize": 20,
        "fontWeight": 900,
      }
    }
  >
    Standard component with two levels
  </span>
</div>
`;

exports[`useSlots sample code test suite renders sample 3 - the two types of higher order header components 1`] = `
<div>
  <span>
    --- SIMPLE USAGE COMPARISON ---
  </span>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Standard HOC
    </span>
  </div>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Staged HOC
    </span>
  </div>
  <span>
    --- COMPARISON WITH CAPTIONS ---
  </span>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Standard HOC with Caption
    </span>
    <span
      style={
        {
          "fontWeight": 900,
        }
      }
    >
      Caption text
    </span>
  </div>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Staged HOC with Caption
    </span>
    <span
      style={
        {
          "fontWeight": 900,
        }
      }
    >
      Caption text
    </span>
  </div>
  <span>
    --- COMPARISON WITH CAPTIONS AND CUSTOMIZATIONS ---
  </span>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "color": "red",
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Standard HOC with caption and customizations
    </span>
    <span
      style={
        {
          "color": "yellow",
          "fontWeight": 900,
        }
      }
    >
      Caption text
    </span>
  </div>
  <div
    style={
      {
        "backgroundColor": "gray",
        "borderColor": "purple",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <span
      style={
        {
          "color": "red",
          "fontSize": 20,
          "fontWeight": 900,
        }
      }
    >
      Staged HOC with caption and customizations
    </span>
    <span
      style={
        {
          "color": "yellow",
          "fontWeight": 900,
        }
      }
    >
      Caption text
    </span>
  </div>
</div>
`;
